runtime/compiler/trj9/runtime/J9Profiler.cpp

```diff
@@ -2602,3 +2618,236 @@ TR_PersistentProfileInfo *TR_AccessedProfileInfo::get(TR_ResolvedMethod *vmMetho
+                                                                &jProfilerThreadProc,
+                                                                javaVM->jitConfig,
+                                                                J9THREAD_CATEGORY_SYSTEM_JIT_THREAD))
+         {
+         if (TR::Options::getVerboseOption(TR_VerboseProfiling))
+            TR_VerboseLog::writeLineLocked(TR_Vlog_PROFILING, "Failed to start jProfiler thread");
+
+         // Thread creation failed, don't try again
+         TR::Options::getCmdLineOptions()->setOption(TR_DisableJProfilerThread);
+         _jProfilerMonitor = NULL;
+         }
+      else
+         {
+         // Wait until creation completes
+         _jProfilerMonitor->enter();
+         while (_state == Initial)
+            _jProfilerMonitor->wait();
+         _jProfilerMonitor->exit();
+
+         if (TR::Options::getVerboseOption(TR_VerboseProfiling))
+            TR_VerboseLog::writeLineLocked(TR_Vlog_PROFILING, "Started jProfiler thread");
+         }
+      }
+   else
+      {
+      if (TR::Options::getVerboseOption(TR_VerboseProfiling))
+         TR_VerboseLog::writeLineLocked(TR_Vlog_PROFILING, "Failed to create JIT-jProfilerMonitor");
+      }
+   }
+
+/**
+ * Stop the profiler thread.
+ * This will set the state to Stop and wait.
+ */
+void
+TR_JProfilerThread::stop(J9JavaVM *javaVM)
+   {
+   PORT_ACCESS_FROM_JAVAVM(javaVM);
+   if (!_jProfilerMonitor)
+      return;
+
+   // Check the thread actually started
+   _jProfilerMonitor->enter();
+   if (!getJProfilerThread())
+      {
+      _jProfilerMonitor->exit();
+      return;
+      }
+
+   if (TR::Options::getVerboseOption(TR_VerboseProfiling))
+      TR_VerboseLog::writeLineLocked(TR_Vlog_PROFILING, "Stopping jProfiler thread");
+
+   // Set state and wait for it to stop
+   _state = Stop;
+   while (getJProfilerThread())
+      {
+      _jProfilerMonitor->notifyAll();
+      _jProfilerMonitor->wait();
+      }
+
+   if (TR::Options::getVerboseOption(TR_VerboseProfiling))
+      TR_VerboseLog::writeLineLocked(TR_Vlog_PROFILING, "Stopped jProfiler thread");
+
+   _jProfilerMonitor->exit();
+   }
+
+/**
+ * Process the persistent profile info.
+ * Should only be called in the profiler thread.
+ */
+void
+TR_JProfilerThread::processWorkingQueue()
+   {
+   while (_state == Run)
+      {
+      _jProfilerMonitor->enter();
+
+      _jProfilerMonitor->wait_timed(_waitMillis, 0);
+      if (_state == Stop)
+         {
+         _jProfilerMonitor->exit();
+         break;
+         }
+
+      _jProfilerMonitor->exit();
+
+      TR_PersistentProfileInfo **prevPtr = &_listHead;
+      TR_PersistentProfileInfo *cur = _listHead;
+
+      size_t index = 0;
+      while (cur && _state == Run)
+         {
+         if (cur->_refCount == 0)
+            {
+            // Clear out entries that are no longer referenced
+            cur = deleteProfileInfo(prevPtr, cur);
+            continue;
```

Comment:
Good point. The atomic compare and exchange on delete will prevent the double delete from happening, as `_listHead` will not match the value it expects. However, it will never correct for this, instead failing the CAS forever. This has been corrected by giving up if the first attempt fails, as a future pass will most likely get it.